crowbar_framework/lib/crowbar/repository.rb

```diff
@@ -48,109 +82,197 @@ def all_platforms
-        repomd_key_md5 == md5
-      else
-        false
-      end
+    def exist?
+      remote? || check_directory
     end
 
-    def repodata_path
-      "/srv/tftpboot/#{@platform}/repos/#{@registry['name']}/repodata"
+    def valid_repo?
+      remote? || check_repo_tag
+    end
+
+    def valid_key_file?
+      remote? || check_key_file
     end
 
-    def repomd_key_md5
-      @registry["repomd"]["md5"]
+    def name
+      @config["name"] || @id
     end
 
-    def repomd_key_path
-      @registry["repomd"]["key"] || "#{repodata_path}/repomd.xml.key"
+    def required
+      @config["required"] || "optional"
     end
 
     def url
-      @registry["url"] || \
-        "http://#{Repository.admin_ip}:#{Repository.web_port}/#{@platform}/repos/#{@registry['name']}"
+      @config["url"] || \
+        "http://#{Repository.admin_ip}:#{Repository.web_port}/#{@platform}/repos/#{name}/"
     end
 
     def active?
-      !bag_item.nil?
+      active_repos = Chef::DataBag.load("crowbar/repositories") rescue {}
+      active_repos.fetch(@platform, {}).key? @id
     end
 
-    def to_databag!
-      repository_item = Chef::DataBagItem.new
-      repository_item.data_bag "repositories"
-      repository_item["id"] = @id
-      repository_item["platform"] = @platform
-      repository_item["name"] = @registry["name"]
-      repository_item["url"] = url
-      repository_item["product_name"] = @registry["product_name"]
-      repository_item
+    def stale?
+      active_repos = Chef::DataBag.load("crowbar/repositories") rescue {}
+      if active_repos.fetch(@platform, {}).key? @id
+        active_repos[@platform][@id] != to_databag_hash
+      else
+        false
+      end
     end
 
-    def bag_item
-      Chef::DataBagItem.load("repositories", @id) rescue nil
+    def to_databag_hash
+      item = Hash.new
+      item["name"] = name
+      item["url"] = url
+      item["ask_on_error"] = @config["ask_on_error"] || false
+      item
     end
 
     private
 
+    def repos_path
+      Pathname.new(File.join("/srv/tftpboot", @platform, "repos"))
+    end
+
+    def repodata_path
+      repos_path.join(name, "repodata")
+    end
+
+    def repodata_media_path
+      repos_path.join(name, "suse", "repodata")
+    end
+
     #
     # validation helpers
     #
-    def all_repo_dirs
-      path = File.join("/srv/tftpboot", @platform, "repos")
-      Dir["#{path}/*"].map { |p| p.split("/").last }
+    def check_directory
+      repos_path.join(name).directory?
     end
 
     def check_repo_tag
-      expected = @registry["repomd"]["tag"]
-      repomd_path = "#{repodata_path}/repomd.xml"
-      if File.exist?(repomd_path)
-        REXML::Document.new(File.open(repomd_path)).root.elements["tags/repo"].text == expected
+      expected = @config.fetch("repomd", {})["tag"]
+      return true if expected.blank?
+
+      repomd_path = repodata_path.join("repomd.xml")
+      repomd_path = repodata_media_path.join("repomd.xml") unless repomd_path.exist?
```

Comment:
`repomd_path ||=` is totally different. It's setting it only if the file exists.